### PR TITLE
Extract independent build safeguards from Phase 2

### PR DESCRIPTION
**Pull request by alexmojaki**
## Summary

- clarify the Python runtime documentation for conditional pricing without regenerating the frozen v2 schemas
- run the existing v2 schema-freeze check for merge-queue candidates against the queue's base commit
- require usage, price, and dimension keys to be ASCII public identifiers that avoid one shared Python/JavaScript/Go blacklist

## Scope

These changes are independent of the v3 publication and runtime-activation work in #672. This PR deliberately excludes background-updater lifecycle changes because #404 defines the desired behavior.

The proposed slim-v2 characterization test was also omitted because `main` already contains the equivalent `test_v2_slim_payload_is_exact_projection_of_full_payload` from #670.

The JavaScript runtime edge-case changes and the Go declaration scanner were removed in favor of the central build-time key policy.

## Validation

- `GOTOOLCHAIN=go1.25.0 make all`
  - Python: 1,548 passed, 39 xfailed, 100% coverage
  - JavaScript: 2,027 passed, 1 skipped
  - Go: race suite passed, 92.9% statement coverage

### CODE REVIEW

File: prices/src/prices/go_identifiers.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    brace_depth = 0
+    source_lines = source.splitlines()
+    code_lines = _go_source_without_comments_and_literals(source).splitlines()
+    for line_index, line in enumerate(code_lines):
+        stripped = line.strip()
+        if declaration_block_kind is not None:
+            if brace_depth == 0 and stripped == ')':
+                declaration_block_kind = None
+                continue
+            if brace_depth == 0 and (match := _GO_BLOCK_MEMBER_PATTERN.match(stripped)):
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply prices/src/prices/go_identifiers.py]: (AI) Addressed in 0e0ffb9c. The grouped-declaration scanner now tracks nested parentheses as well as braces, and a regression covers multiline function signatures in both `type` and `var` blocks.

File: prices/src/prices/export_validation.py [hunk 2]
```diff
@@ -119,13 +119,15 @@ def validate_units(raw_units: Mapping[str, Mapping[str, Any]]) -> UnitRegistry:
...
     """Validate registry structure, provider model prices, and provider extractors before export."""
+    from prices.go_identifiers import validate_go_usage_key_identifiers
     from prices.package_data import (
         validate_provider_extractor_destinations,
         validate_provider_extractor_reasoning_coverage,
         validate_provider_model_prices,
     )
 
     registry = validate_units(units)
+    validate_go_usage_key_identifiers(registry.units)
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply prices/src/prices/export_validation.py]: (AI) Addressed in 6d19d3e0. `package_go_data()` now validates the usage-key vocabulary before creating or writing any Go package artifacts, with a regression covering direct packaging.

File: packages/js/src/__tests__/units.test.ts [hunk 3]
```diff
@@ -196,6 +196,41 @@ describe('UnitRegistry', () => {
     )
   })
 
+  it('uses locale-independent dimension ordering for joins', () => {
+    const registry = new UnitRegistry({
+      composed_events: { dimensions: { Å: 'ring', Å: 'precomposed', family: 'events' }, per: 1 },
```
Comment (alexmojaki) [inline packages/js/src/__tests__/units.test.ts]: stop spending energy on weird edge cases about units. require them to be ascii in build, don't worry about this at runtime.

File: prices/src/prices/package_data.py [hunk 4]
```diff
@@ -238,6 +239,7 @@ def _runtime_unit_data(units: Mapping[str, Mapping[str, Any]]) -> dict[str, dict
 
 def package_go_data(provider_data: JsonData, units: Mapping[str, Mapping[str, Any]]) -> None:
     """Generate the data bundled by the Go package."""
+    validate_go_usage_key_identifiers(units)
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply prices/src/prices/package_data.py]: (AI) Addressed in 199961d6. Identifier validation now runs in `package_data()` before any Python, Go, or TypeScript generator is called; the regression asserts generation never starts for invalid input.

File: prices/src/prices/go_identifiers.py
Comment (alexmojaki) [inline prices/src/prices/go_identifiers.py:1]: why so much complication? can we not just have one blacklist of unit names that covers all languages? that's already very cautious. we don't need to be paranoid that someone will try their best to find a unit name that breaks things.

File: prices/src/prices/go_identifiers.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            brace_depth += line.count('{') - line.count('}')
+            parenthesis_depth += line.count('(') - line.count(')')
+            continue
+
+        if brace_depth != 0:
+            brace_depth += line.count('{') - line.count('}')
+            continue
+        if match := _GO_BLOCK_PATTERN.fullmatch(stripped):
+            declaration_block_kind = match.group(1)
+        elif match := _GO_DECLARATION_PATTERN.match(stripped):
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply prices/src/prices/go_identifiers.py]: (AI) Addressed in b1b9fd30. The Python regex scanner has been replaced with Go's native AST parser. The regression covers multiline generic functions, comments between declaration tokens, grouped multiline function types, continued expressions, local variables, methods, generated constants, and files from other packages.

File: prices/src/prices/go_identifiers.py
Comment (alexmojaki) [reply prices/src/prices/go_identifiers.py:1]: (AI) Addressed in a0f09efc and 807c1e69. The Go declaration scanner and helper are gone. Build validation now uses one shared reserved-key blacklist covering Python, JavaScript, and Go, including `key` for the generated Go `UsageKey` collision.

File: packages/js/src/__tests__/units.test.ts (hunk 3 above)
Comment (alexmojaki) [reply packages/js/src/__tests__/units.test.ts]: (AI) Addressed in a0f09efc and 807c1e69. The JavaScript Unicode/prototype runtime changes and tests are gone. Usage, price, and dimension keys must now be ASCII public identifiers and avoid the shared build-time blacklist.

File: prices/src/prices/export_validation.py [hunk 6]
```diff
@@ -92,6 +107,8 @@ def validate_units(raw_units: Mapping[str, Mapping[str, Any]]) -> UnitRegistry:
             raise ValueError(f'Invalid per for unit {usage_key}: expected a positive integer, got {per!r}')
 
         dimensions = dict(cast(Mapping[str, str], raw_unit.get('dimensions', {})))
+        for dimension_key in dimensions:
+            _validate_public_key('dimension', dimension_key)
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply prices/src/prices/export_validation.py]: (AI) Addressed in 31e018a4. Build validation now rejects non-string dimension keys with the same `ValueError` diagnostic used for other non-public keys, with a focused regression.
